src/factory.ts

```diff
@@ -86,23 +94,54 @@ export function buildDependencies(
     return output
 }
 
+/**
+ * Set the maintain values and container sizes for each container in the factory
+ * @param factory FactoryGraph to update
+ */
+export function updateContainers(factory: FactoryGraph) {
```

Comment:
I would prefer this to be computed in getter functions on the `ContainerNode` itself. If we make any changes to the graph (e.g. by pruning unnecessary machines), we would have to call this function again. 

Also, the type of `maintain` wouldn't need to include `undefined` in case of a getter. (However, this could also be achieved by initializing the field with 0.)

```typescript
class ContainerNode {
    get maintain(): Quantity {
        let maintain = 0
        for (const consumer of this.consumers) {
            for (const ingredient of findRecipe(consumer.item).ingredients) {
                if (ingredient.item === this.item) {
                    maintain += ingredient.quantity
                }
            }
            /* If consumer is an OutputNode, maintain OutputNode requirement */
            if (isOutputNode(consumer) && consumer.item === this.item) {
                maintain += consumer.maintain
            }
        }
        return maintain
    }

    get size(): ContainerSize | undefined {
        const maintainVolume = this.maintain * this.item.volume
        let constainerSize: ContainerSize | undefined
        for (const [size, volume] of ContainerVolume) {
            if (maintainVolume < volume) {
                containerSize = size
            }
        }
        return containerSize
    }
}

Reply:
Done in https://github.com/lgfrbcsgo/du-factory-generator/commit/748b5b78bc4463229ba41a097321bc7fb77e6303